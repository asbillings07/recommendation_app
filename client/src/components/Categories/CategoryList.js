import React, { Component } from 'react';
import Axios from 'axios';
import Config from '../../Config';
import { Link } from 'react-router-dom';
import { Container, Button, Row, Col } from 'react-bootstrap';
import styled from 'styled-components';

class CategoryList extends Component {
  state = {
    categories: [],
    loading: true,
  };

  componentDidMount() {
    this.getAllCategories();
  }

  getAllCategories = async () => {
    try {
      const categories = await Axios.get(`${Config.apiBaseUrl}/category`).catch(
        err => console.log(err)
      );
      if (categories) {
        console.log(categories.data.category);
        this.setState({ categories: categories.data.category, loading: false });
      }
    } catch (err) {
      console.log(err);
      this.props.history.push('/notfound');
    }
  };

  showCategories = () => {
    return this.state.categories.map(category => (
      <Col className="mb-2" sm={4} key={category.id}>
        <Link to="#">
          <Button variant="primary" size="lg" block>
            {category.title}
          </Button>
        </Link>
      </Col>
    ));
  };

  render() {
    return (
      <React.Fragment>
        <H2>Browse Categories</H2>
        <Container className="mb-7 mt-3">
          <Row>{this.showCategories()}</Row>
        </Container>
      </React.Fragment>
    );
  }
}

export default CategoryList;

const H2 = styled.h2`
  display: flex;
  justify-content: center;
  font-family: 'Oswald', sans-serif;
`;
