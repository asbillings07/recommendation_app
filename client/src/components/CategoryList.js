import React, { Component } from 'react'

class CategoryList extends Component {
    render() {

        let categoryMockArray = this.props.categories.categories
        let allCategories = categoryMockArray.map(singleCategory => {
            return (<div>
                <p>{singleCategory.title}</p>
            </div>)
        })
        return (
            <div>
                {allCategories}
            </div>
        )
    }
}


export default CategoryList;