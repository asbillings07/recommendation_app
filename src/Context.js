import React, { useState } from 'react'
import Data from './Data'
import Cookies from 'js-cookie'
export const Context = React.createContext()

/**
 * function that Provides context
 * SignIn and SignOut Methods
 *
 */

export const Provider = ({ children }) => {
  const data = new Data()

  const initalAuthState = Cookies.getJSON('authorizedUser') || null
  const initalTokenState = Cookies.getJSON('token') || null
  const [state, setState] = useState({
    authorizedUser: initalAuthState,
    token: initalTokenState
  })

  const signIn = async (email, password) => {
    const creds = { email, password }
    const user = await data.login(creds)
    // console.log(user)
    if (user) {
      setState({ authorizedUser: user.user, token: user.token })
      Cookies.set('authorizedUser', JSON.stringify(user.user), { expires: 1 })
      Cookies.set('token', JSON.stringify(user.token), { expires: 1 })
    }
    return user
  }

  /** SignOut Method - Signs out user and removes cookies */

  const signOut = () => {
    setState({ authorizedUser: null, token: null })
    Cookies.remove('authorizedUser')
    Cookies.remove('token')
    console.log('SignOut Successful')
  }

  const value = {
    authorizedUser: state.authorizedUser,
    token: state.token,
    data,
    actions: {
      signIn,
      signOut
    }
  }

  return <Context.Provider value={value}>{children}</Context.Provider>
}

export const Consumer = Context.Consumer

/**
 * A higher-order component that wraps the provided component in a Context Consumer component.
 * @param {class} Component - A React component.
 * @returns {function} A higher-order component.
 */

export default function withContext(Component) {
  return function ContextComponent(props) {
    return (
      <Context.Consumer>{(context) => <Component {...props} context={context} />}</Context.Consumer>
    )
  }
}
