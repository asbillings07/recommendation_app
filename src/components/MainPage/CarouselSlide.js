import React from "react";
import { Carousel } from "react-bootstrap";
import city3 from "../../images/city3.jpg";
import city4 from "../../images/city4.jpg";
import city5 from "../../images/city5.jpg";
import styled from "styled-components";

export default function CarouselSlide() {
  return (
    <Carousel aria-label="carousel slide" className="mb-5">
      <Carousel.Item>
        <StyledImg src={city5} alt="First slide" />
        <Carousel.Caption aria-label="first slide caption says">
          <h3>Welcome to RecommendIt!</h3>
        </Carousel.Caption>
      </Carousel.Item>
      <Carousel.Item>
        <StyledImg src={city3} alt="Second slide" />

        <Carousel.Caption aria-label="second slide caption says">
          <h3>Browse categories for recommendations or create your own!</h3>
        </Carousel.Caption>
      </Carousel.Item>
      <Carousel.Item>
        <StyledImg src={city4} alt="Third slide" />

        <Carousel.Caption aria-label="third slide caption says">
          <h3>Comment on recommendations!</h3>
        </Carousel.Caption>
      </Carousel.Item>
    </Carousel>
  );
}

const StyledImg = styled.img`
  display: block;
  width: 100vw;
  max-height: 50vh;
  overflow: hidden;
`;
