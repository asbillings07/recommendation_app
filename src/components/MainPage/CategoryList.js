import React from 'react'
import { Link } from 'react-router-dom'
import { Container, Button, Row, Col } from 'react-bootstrap'
import SearchBar from './SearchBar'
import styled from 'styled-components'
import Spinner from '../Spinner'

export default function CategoryList({ data, categories, setCategories, loading }) {
  const showCategories = () => {
    return categories.map((category) => (
      <Col className='mb-2' sm={4} key={category.id}>
        <StyledLink data-testid='category-link' to={`/category/${category.id}/recs`}>
          <CategoryButton
            aria-label={`navigate to ${category.title}`}
            variant='outline-primary'
            size='lg'
            block
          >
            {category.title}
          </CategoryButton>
        </StyledLink>
      </Col>
    ))
  }

  if (loading) {
    return <Spinner size='4x' spinning='spinning' />
  } else {
    return (
      <>
        <SearchBar categories={data} setCategories={setCategories} />
        <H2 aria-label='Browse Categories'>Browse Categories</H2>
        <CategoryContainer>
          <Row>{showCategories()}</Row>
        </CategoryContainer>
      </>
    )
  }
}

const H2 = styled.h2`
  display: flex;
  justify-content: center;
  font-family: 'Oswald', sans-serif;
`
const CategoryContainer = styled(Container)`
  margin-bottom: 9px;
  margin-top: 3px;
`
const CategoryButton = styled(Button)`
  outline-color: #1168d9 !important;
  margin-bottom: 0.25em;

  :hover {
    background-color: #1168d9 !important;
    box-shadow: 0px 11px 13px -2px rgba(32, 44, 121, 0.65), 0px 20px 20px -26px rgb(0, 0, 0);
  }
`
const StyledLink = styled(Link)`
  text-decoration: none;
  :hover {
    text-decoration: none !important;
  }
`
