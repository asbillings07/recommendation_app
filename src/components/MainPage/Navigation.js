import React from 'react'
import { Navbar, Nav } from 'react-bootstrap'
import { useSelector } from 'react-redux'
import styled from 'styled-components'

const Navigation = () => {
  const { authorizedUser } = useSelector((state) => state.users)
  return (
    <NavDiv>
      <NavBar collapseOnSelect expand='lg'>
        <NavBarTitle aria-label='App Title is RecommendIt' href='/'>
          Recommend It!
        </NavBarTitle>
        <Navbar.Toggle aria-controls='responsive-navbar-nav' />
        <Navbar.Collapse id='responsive-navbar-nav'>
          {authorizedUser ? (
            <>
              <Nav className='mr-auto'></Nav>
              <Nav variant='pills'>
                <NavLink
                  aria-label={`Welcome, ${authorizedUser.firstName}!`}
                  user='true'
                >{`Welcome, ${authorizedUser.firstName}!`}</NavLink>
                <Nav.Item>
                  <NavLink aria-label='This link displays the Profile page' href='/profile'>
                    Profile
                  </NavLink>
                </Nav.Item>
                <Nav.Item>
                  <NavLink aria-label='This link logs you out' href='/signout'>
                    Sign Out
                  </NavLink>
                </Nav.Item>
              </Nav>
            </>
          ) : (
            <>
              <Nav className='mr-auto'></Nav>
              <Nav>
                <NavLink aria-label='Welcome, User!' user='true'>
                  Welcome, User!
                </NavLink>
                <NavLink aria-label='This link displays the signIn page' href='/signin'>
                  SignIn
                </NavLink>
                <NavLink aria-label='This link displays the signup page' href='/signup'>
                  SignUp
                </NavLink>
              </Nav>
            </>
          )}
        </Navbar.Collapse>
      </NavBar>
    </NavDiv>
  )
}

export default Navigation

const NavDiv = styled.div`
  margin-bottom: 0px;
`
const NavBar = styled(Navbar)`
  background-color: #f25c05 !important;
`
const NavBarTitle = styled(NavBar.Brand)`
  color: white !important;
`
const NavLink = styled(Nav.Link)`
  color: ${(props) => (props.user ? 'black' : 'white')} !important;
`
