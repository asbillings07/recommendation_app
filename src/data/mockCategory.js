export const categories = [
  {
    id: 1,
    title: 'Restaurants',
  },
  {
    id: 2,
    title: 'Beauty & Spas',
  },
  {
    id: 3,
    title: 'Automotive',
  },
  {
    id: 4,
    title: 'Education',
  },
  {
    id: 5,
    title: 'Pets',
  },
  {
    id: 6,
    title: 'Events',
  },
  {
    id: 7,
    title: 'Home Services',
  },
  {
    id: 8,
    title: 'Daycare',
  },
  {
    id: 9,
    title: 'Arts & Entertainment',
  },
  {
    id: 10,
    title: 'Real Estate',
  },
  {
    id: 11,
    title: 'Hotels',
  },
  {
    id: 12,
    title: 'Outdoor Recreation',
  },
  {
    id: 13,
    title: 'Health and Medical',
  },
  {
    id: 14,
    title: 'Nightlife',
  },
  {
    id: 15,
    title: 'Shopping',
  },
];
