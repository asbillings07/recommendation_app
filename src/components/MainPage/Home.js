/* eslint-disable react-hooks/exhaustive-deps */
import React, { useState, useEffect } from 'react'
import CategoryList from './CategoryList'
import Axios from 'axios'
import Config from '../../Config'
import CarouselSlide from './CarouselSlide'

export default function Home({ context, history }) {
  const [data, setData] = useState([])
  const [loading, setLoading] = useState(true)
  const [categories, setCategories] = useState([])
  const env = Config.env
  const getAllCategories = async () => {
    try {
      const categories = await Axios.get(`${Config[env].apiBaseUrl}/category`).catch(err =>
        console.log(err)
      )
      if (categories) {
        setData(categories.data.category)
        setLoading(false)
      }
    } catch (err) {
      console.log(err)
      history.push('/notfound')
    }
  }

  useEffect(() => {
    getAllCategories()
  }, [context, history])

  return (
    <>
      <CarouselSlide authUser={context.authorizedUser} />
      <CategoryList
        data={data}
        categories={categories}
        setCategories={setCategories}
        loading={loading}
      />
    </>
  )
}
