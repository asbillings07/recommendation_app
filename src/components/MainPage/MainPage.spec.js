import React from 'react';
import { render, cleanup } from '@testing-library/react';
import { Provider } from '../../Context';
import Home from './Home';
import CarouselSlide from './CarouselSlide';
import Navigation from './Navigation';
import CategoryList from './CategoryList';
import SearchBar from './SearchBar';
import { MemoryRouter } from 'react-router-dom';
import { categories } from '../../data/mockCategory';

global.console.error = jest.fn();

describe('All components in on the Main(landing) page render correctly', () => {
  afterEach(cleanup);
  const context = {
    authorizedUser: {
      firstName: 'Aaron',
      lastName: 'Billings',
    },
  };

  const category = [
    {
      id: 1,
      title: 'Resturant',
      Recommendations: [],
    },
    {
      id: 2,
      title: 'Outdoor Activities',
      Recommendations: [],
    },
    {
      id: 3,
      title: 'Automotive',
      Recommendations: [],
    },
  ];

  test('<Home/>', () => {
    const { container } = render(
      <Provider>
        <Home context={context} />
      </Provider>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test('<CarouselSlide/>', () => {
    const { queryAllByText } = render(<CarouselSlide />);

    expect(queryAllByText('first slide caption says')).toBeTruthy();
    expect(queryAllByText('second slide caption says')).toBeTruthy();
    expect(queryAllByText('third slide caption says')).toBeTruthy();
  });

  test('<SearchBar/>', () => {
    const setCategories = jest.fn();

    const { debug, getByPlaceholderText } = render(
      <SearchBar setCategories={setCategories} />
    );

    expect(getByPlaceholderText('Search')).toBeTruthy();
  });

  test('<Navigation/>', () => {
    const { queryAllByLabelText } = render(<Navigation context={context} />);
    expect(queryAllByLabelText('App Title is RecommendIt')).toBeTruthy();
    expect(
      queryAllByLabelText(`Welcome, ${context.authorizedUser.firstName}!`)
    ).toBeTruthy();
    expect(
      queryAllByLabelText('This link displays the Profile page')
    ).toBeTruthy();
    expect(queryAllByLabelText('This link logs you out')).toBeTruthy();
  });

  test('<CategoryList/>', () => {
    const setCategories = jest.fn();
    const { queryByLabelText, getAllByTestId } = render(
      <MemoryRouter>
        <CategoryList setCategories={setCategories} categories={categories} />
      </MemoryRouter>
    );

    expect(queryByLabelText(`navigate to ${category.title}`));
    expect(getAllByTestId('category-link')).toBeTruthy();
  });
});
