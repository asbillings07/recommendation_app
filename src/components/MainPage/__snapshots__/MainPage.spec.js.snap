// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`All components in on the Main(landing) page render correctly <Home/> 1`] = `
<div
  aria-label="carousel slide"
  class="mb-5 carousel slide"
>
  <ol
    class="carousel-indicators"
  >
    <li
      class="active"
    />
    <li />
    <li />
  </ol>
  <div
    class="carousel-inner"
  >
    <div
      class="active carousel-item"
    >
      <img
        alt="First slide"
        class="sc-bZQynM fyaQNJ"
        src="city5.jpg"
      />
      <div
        aria-label="first slide caption says"
        class="carousel-caption"
      >
        <h3>
          Welcome to RecommendIt!
        </h3>
      </div>
    </div>
    <div
      class="carousel-item"
    >
      <img
        alt="Second slide"
        class="sc-bZQynM fyaQNJ"
        src="city3.jpg"
      />
      <div
        aria-label="second slide caption says"
        class="carousel-caption"
      >
        <h3>
          Browse categories for recommendations or create your own!
        </h3>
      </div>
    </div>
    <div
      class="carousel-item"
    >
      <img
        alt="Third slide"
        class="sc-bZQynM fyaQNJ"
        src="city4.jpg"
      />
      <div
        aria-label="third slide caption says"
        class="carousel-caption"
      >
        <h3>
          Comment on recommendations!
        </h3>
      </div>
    </div>
  </div>
  <a
    class="carousel-control-prev"
    href="#"
    role="button"
  >
    <span
      aria-hidden="true"
      class="carousel-control-prev-icon"
    />
    <span
      class="sr-only"
    >
      Previous
    </span>
  </a>
  <a
    class="carousel-control-next"
    href="#"
    role="button"
  >
    <span
      aria-hidden="true"
      class="carousel-control-next-icon"
    />
    <span
      class="sr-only"
    >
      Next
    </span>
  </a>
</div>
`;
